### PR TITLE
Search resources

### LINKED ISSUE #296: [FEATURE] Build a reusable search component, similar to a modularized filter component
**Describe your suggested improvement**
It would be nice to add a traditional search bar to filter resources and videos (rather than only filter on tags/channels). I am open to suggestions on what fields should be searchable. Also, we can start this functionality on the front-end but eventually the searching could happen on the backend and we can build this to make sure that transition is "seamless".

**Describe the benefit**
Users are more easily able to find relevant resources. 

**Would you be capable/willing to implement the improvement?**
Yes but I might not have time this week

**Additional comments**
@VeckoTheGecko let me know if you have any thoughts on how this should interact with the data.

### PR DESCRIPTION
**Pull request by christina-ml**
## Describe your changes
Added search bar filter and moved it into its own component to filter resources

## Related issue number/link
Fixes #296

Thanks for contributing! 🥳🚀🌳
Screenshot 2023-08-01 at 4 15 41 PM

### CODE REVIEW

File: src/routes/resources/+page.svelte [hunk 1]
```diff
@@ -55,21 +82,29 @@
...
     displayedResources = displayedResources;
   }
 
+  // Update the displayedResourceLimit based on the scroll position
   const updateLimit = (event: CustomEvent<{displayLimit: number}>) => {
     const {displayLimit} = event.detail
     displayedResourceLimit = displayLimit
   }
+
+  let mainH1El: HTMLHeadingElement | null;
```
Comment (dmlb) [inline src/routes/resources/+page.svelte:98]: some of these look like artifacts possibly from merge conflicts; is this in use in the file?

File: src/lib/components/Search.svelte [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    const dispatch = createEventDispatcher();
+
+    function onSubmit(){
+        dispatch("search", {searchTerm})
+    }
+
+</script>
+
+<form on:submit|preventDefault={onSubmit} class="p-4 space-y-4">
```
Comment (dmlb) [inline src/lib/components/Search.svelte:14]: role search seems appropriate here https://www.w3.org/WAI/ARIA/apg/patterns/landmarks/examples/search.html

File: src/lib/components/Search.svelte [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	<label for="search" />
+	<input
+		type="text"
+		bind:value={searchTerm}
+		placeholder="Search by keyword"
+		id="search"
+		name="search"
+		class="p-2 rounded-lg border-2 border-green-500 dark:border-green-700 text-green-700 dark:text-green-500"
```
Comment (dmlb) [inline src/lib/components/Search.svelte:22]: it is missing a background colour for dark mode and doesn't align left with the rest of the content:
Screen Shot 2023-08-02 at 8 58 39 PM

File: src/lib/components/Search.svelte [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	/>
+	<button
+		type="submit"
+		class="p-2 rounded-lg bg-green-700 text-white dark:bg-green-900/75"
+	>
+		Search
+	</button>
```
Comment (dmlb) [inline src/lib/components/Search.svelte:29]: should be able to use the component here for a single source of truth on all these colour classes for light/dark/hover


```suggestion
	 <ButtonLinks
                    type="submit"
                    version="filled"
                    color="green"
                >
                    <span slot="label">Search</span>
                </ButtonLinks>
```

File: src/routes/resources/+page.svelte [hunk 5]
```diff
@@ -2,8 +2,10 @@
   import type { PageData } from "./$types";
   import { onMount } from "svelte";
   import { DEFAULT_DISPLAY_LIMIT } from "$lib/constants";
-  import type { Tag, FilterOption, FilterLogic } from "$lib/interfaces";
+  import type { Tag, FilterOption, FilterLogic, Resource } from "$lib/interfaces";
   import { setIntersection } from "$lib/utils";
+  import Collapsible from "$lib/components/Collapsible.svelte";
+  import Search from "lib//components/Search.svelte";
```
Comment (dmlb) [inline src/routes/resources/+page.svelte:8]: `lib//` vs `$lib/`
and not sure Collapsible is in this file?

**Comment by VeckoTheGecko** [comment]
Hi @christina-ml , thanks for the PR! Would you be able to provide some insight into how the implemented search functionality differs from fuzzy searching packages like Fuse.js? (https://www.fusejs.io/) Would fusejs be a quick, free way to offload some code and get some good search functionality for the codebase in return?

File: src/routes/resources/+page.svelte (hunk 1 above)
Comment (christina-ml) [reply src/routes/resources/+page.svelte:98]: Correct, this line of code is not in use and can be removed.

File: src/routes/resources/+page.svelte (hunk 5 above)
Comment (christina-ml) [reply src/routes/resources/+page.svelte:8]: Collapsible has been removed from this file, so there is no need to import it here.
lib// vs $lib/ appears to work either way, but changing it to $lib/ to keep it consistent with the other imports.

File: src/lib/components/Search.svelte (hunk 3 above)
Comment (christina-ml) [reply src/lib/components/Search.svelte:22]: Add background color from white to transparent. Align form to the left by removing padding and adding margin to the bottom.

File: src/lib/components/Search.svelte (hunk 2 above)
Comment (christina-ml) [reply src/lib/components/Search.svelte]: Add role search to form and input tags for Search.svelte

**Comment by christina-ml** [comment]
@VeckoTheGecko Added fuse.js to implement fuzzy searching. 
Installed fuse.js with command `npm install fuse.js`
Set the threshold to 0.25 which matches close enough to most searches I tested.
Screenshot 2023-08-15 at 4 34 44 PM

File: src/lib/components/Search.svelte (hunk 4 above)
Comment (christina-ml) [reply src/lib/components/Search.svelte]: Using ButtonLinks component for light/dark/hover

**Comment by christina-ml** [comment]
@VeckoTheGecko I have resolved all above changes on this pull request.

**Review by dmlb — APPROVED** [review]

File: src/routes/resources/+page.svelte [hunk 6]
```diff
@@ -30,8 +36,29 @@
     filterObject.push(tagOption)
   }
 
+  function filterBySearchInput(event: CustomEvent<{searchTerm: string}>) {
+    const { searchTerm } = event.detail;
+
+    const options = {
+      includeScore: true,
+      threshold: 0.25,
+      keys: ['description', 'tags', 'title']
```
Comment (VeckoTheGecko) [inline src/routes/resources/+page.svelte:45]: What is the effect of including "tags" in `keys`. Is the intention to also search by tags (e.g. by searching "Activism" get to the resources with that tag?*)

I think limiting the search to the titles and summaries would be good

*to be clear, this usecase work from my testing on this branch

**Comment by VeckoTheGecko** [comment]
As a point of discussion: how should the search work when there are also tags being used? At the moment, the search bar and the tag filter are completely independent and don't work in conjunction (which is fine by me). Is there a way that this can be made clearer in the UI? Or is the visual separation between the search bar and the tag filters enough?

**Pull request merged**